File: NSTU-lab3/lab3/src/containers/set.cpp
#ifndef SET_CPP
#define SET_CPP

#include "hash.h"

template<typename T>
Set<T>::Set(size_t initialCapacity) : map(initialCapacity) {}

template<typename T>
void Set<T>::add(const T& value) {
    map.put(value, true);
}

template<typename T>
bool Set<T>::contains(const T& value) const {
    return map.contains(value);
}

template<typename T>
bool Set<T>::remove(const T& value) {
    return map.remove(value);
}

template<typename T>
void Set<T>::clear() {
    map.clear();
}

template<typename T>
size_t Set<T>::size() const {
    return map.size();
}

template<typename T>
bool Set<T>::empty() const {
    return map.empty();
}

template<typename T>
void Set<T>::print(std::ostream& os) const {
    os << "Set { size: " << size() << " }";
}

#endif

File: NSTU-lab3/lab3_go/containers/set.go
package containers

// Set - множество на основе HashMap
type Set struct {
	hashMap *HashMap
}

// NewSet создаёт новое множество
func NewSet(initialCapacity int) *Set {
	return &Set{
		hashMap: NewHashMap(initialCapacity),
	}
}

// Add добавляет элемент в множество
func (s *Set) Add(value string) error {
	return s.hashMap.Put(value, "")
}

// Contains проверяет наличие элемента
func (s *Set) Contains(value string) bool {
	return s.hashMap.Contains(value)
}

// Remove удаляет элемент из множества
func (s *Set) Remove(value string) bool {
	return s.hashMap.Remove(value)
}

// Clear очищает множество
func (s *Set) Clear() {
	s.hashMap.Clear()
}

// Size возвращает количество элементов
func (s *Set) Size() int {
	return s.hashMap.Size()
}

// Print выводит множество
func (s *Set) Print() string {
	return s.hashMap.Print()
}
